modules/Movecraft/src/main/java/net/countercraft/movecraft/async/AsyncTask.java

```diff
@@ -47,4 +54,78 @@ public void run() {
     protected Craft getCraft() {
         return craft;
     }
+
+    protected boolean checkFuel(){
+        // check for fuel, burn some from a furnace if needed. Blocks of coal are supported, in addition to coal and charcoal
+        double fuelBurnRate = craft.getType().getFuelBurnRate(craft.getW());
+        if (fuelBurnRate == 0.0 || craft.getSinking()) {
+            return true;
+        }
+        if (craft.getBurningFuel() >= fuelBurnRate) {
+            double burningFuel = craft.getBurningFuel();
+            //call event
+            final FuelBurnEvent event = new FuelBurnEvent(craft, burningFuel, fuelBurnRate);
+            Bukkit.getPluginManager().callEvent(event);
+            if (event.getBurningFuel() != burningFuel)
+                burningFuel = event.getBurningFuel();
+            if (event.getFuelBurnRate() != fuelBurnRate)
+                fuelBurnRate = event.getFuelBurnRate();
+            craft.setBurningFuel(burningFuel - fuelBurnRate);
+            return true;
+        }
+        Block fuelHolder = null;
+        for (MovecraftLocation bTest : craft.getHitBox()) {
+            Block b = craft.getW().getBlockAt(bTest.getX(), bTest.getY(), bTest.getZ());
+            if (b.getType() == Material.FURNACE) {
+                InventoryHolder inventoryHolder = (InventoryHolder) b.getState();
+                for (Material fuel : craft.getType().getFuelTypes().keySet()) {
```

Comment:
loop over the inventory holder, not the fuel types - one is constant, the other scales